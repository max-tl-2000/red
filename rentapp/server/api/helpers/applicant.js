/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import omit from 'lodash/omit';
import pick from 'lodash/pick';
import { getPersonApplicationToken } from '../../services/person-application';
import { getCommonUserByPersonId, getRegistrationToken } from '../../../../auth/server/services/common-user';
import { Routes } from '../../../common/enums/rentapp-types';
import { resolveSubdomainURL } from '../../../../common/helpers/resolve-url';
import { getPersonById } from '../../../../server/services/person';
import { enhanceAppConfigFromPersonMapping } from '../../../../common/helpers/auth';
import loggerModule from '../../../../common/helpers/logger';
const logger = loggerModule.child({ subType: 'helperApplication' });
import { personApplicationProvider } from '../../providers/person-application-provider-integration';

export const getApplicantRegistrationToken = async (req, options = {}) => {
  const { tenantId, tenantDomain, quoteId, partyId, propertyId, personId, personApplicationId, screeningVersion } = options;
  const ctx = { ...req, tenantId };
  const isApplicationPaid = await personApplicationProvider(screeningVersion).hasPaidApplication(req, personId, partyId);

  const commonUser = await getCommonUserByPersonId(ctx, personId);
  if (!(isApplicationPaid && commonUser)) return undefined;

  const applicantData = {
    partyId,
    quoteId,
    personApplicationId,
    propertyId,
    personId,
    tenantDomain,
  };

  return await getRegistrationToken(ctx, commonUser, applicantData);
};

export const getApplicationUrl = async (req, data) => {
  logger.debug({ ctx: req }, 'getApplicationUrl');
  const ctx = req;
  const { userId, personMapping, appId, hasMultipleApplications } = data;
  ctx.tenantId = personMapping.tenantId;

  const { appConfig } = data.settings || {};
  let { applicant, quoteId, partyId, propertyId } = data;
  if (appConfig) {
    partyId = appConfig.partyId;
    quoteId = appConfig.quoteId;
    propertyId = appConfig.propertyId;
    applicant = {
      personApplicationId: appConfig.personApplicationId,
      tenantDomain: appConfig.tenantDomain,
    };
  }

  const applicationTokenObject = {
    personId: personMapping.personId,
    quoteId,
    partyId,
    tenantId: personMapping.tenantId,
    personApplicationId: applicant.personApplicationId,
    propertyId,
    commonUserId: userId,
    tenantDomain: applicant.tenantDomain,
    hasMultipleApplications,
  };

  const applicationToken = await getPersonApplicationToken(ctx, applicationTokenObject);
  return resolveSubdomainURL(`${req.protocol}://${req.hostname}${Routes.additionalInfo}${applicationToken}`, appId);
};

// TODO: CPM-7383 - we need to get the applicant information in the cases we only have the personApplicationId
export const getApplicantData = async ({ authUser }) => {
  const { personMapping, applicant, settings } = authUser;
  if (settings && settings.appConfig) return enhanceAppConfigFromPersonMapping(settings.appConfig, personMapping);

  if (!(personMapping && applicant)) {
    const person = await getPersonById({ tenantId: authUser.tenantId }, authUser.personId);

    if (!(person && person.mergedWith)) return authUser;

    const mergedPerson = await getPersonById({ tenantId: authUser.tenantId }, person.mergedWith);

    return {
      ...authUser,
      personId: mergedPerson.id,
      personName: mergedPerson.preferredName,
    };
  }

  const { personId } = personMapping || {};
  return {
    quoteId: authUser.quoteId,
    partyId: authUser.partyId,
    propertyId: authUser.propertyId,
    personId,
    personName: authUser.name,
    tenantId: authUser.tenantId,
    tenantDomain: applicant.tenantDomain,
  };
};

export const formatApplicationObject = applicantData => {
  const applicantDataFields = [
    'id',
    'personId',
    'partyId',
    'partyApplicationId',
    'applicationData',
    'additionalData',
    'applicantId',
    'isFeeWaived',
    'paymentLink',
  ];
  const applicationDataFieldsToDelete = ['guarantors', 'invitedToApply', 'otherApplicants'];
  const applicantionDataFieldsNeededForIncompletedPayment = ['firstName', 'middleName', 'lastName'];
  const applicantDataToRemove = ['additionalData'];

  const { paymentCompleted, applicationData } = applicantData;
  const formatedApplicationData = omit(applicationData, applicationDataFieldsToDelete);
  applicantData.applicationData = formatedApplicationData;
  const applicantNeededData = pick(applicantData, applicantDataFields);
  if (paymentCompleted) {
    return applicantNeededData;
  }
  const applicationDataWithoutSensitiveContent = pick(formatedApplicationData, applicantionDataFieldsNeededForIncompletedPayment);
  const applicantDataStepOne = omit(applicantNeededData, applicantDataToRemove);
  applicantDataStepOne.applicationData = applicationDataWithoutSensitiveContent;

  return applicantDataStepOne;
};
