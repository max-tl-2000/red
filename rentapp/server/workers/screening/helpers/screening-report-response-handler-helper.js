/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { createSubmissionResponse, getSubmissionResponseBySubmissionRequestId } from '../../../dal/fadv-submission-repo';
import { DALTypes } from '../../../../../common/enums/DALTypes';
import loggerModule from '../../../../../common/helpers/logger';
import { shouldStoreScreeningResponse } from '../screening-helper';
import { ScreeningDecision } from '../../../../../common/enums/applicationTypes';
import { NoRetryError } from '../../../../../server/common/errors';

const logger = loggerModule.child({ subType: 'Application Screening Response Handler' });

const { ScreeningRecommendation } = DALTypes;

export const handleFADVResponseError = async (ctx, rawResponse, error, { screeningRequestId, responseOrigin }) => {
  logger.error({ ctx, fadvError: error, screeningRequestId }, 'handleFADVResponseError');
  if (!screeningRequestId) throw new NoRetryError('screeningRequestId is required');
  const submissionResponse = {
    submissionRequestId: screeningRequestId,
    rawResponse,
    applicationDecision: error,
    origin: responseOrigin,
  };

  const recommendation = { id: 'reva-error' };

  if (error === ScreeningDecision.ERROR_RESPONSE_UNPARSABLE) {
    recommendation.text = ScreeningRecommendation.GENERIC_ERROR_RECOMMENDATION;
  } else {
    submissionResponse.externalId = rawResponse.ApplicantScreening.Response[0].RequestID_Returned[0];
    recommendation.text =
      error === ScreeningDecision.ERROR_ADDRESS_UNPARSABLE
        ? ScreeningRecommendation.ERROR_ADDRESS_UNPARSABLE_RECOMMENDATION
        : ScreeningRecommendation.GENERIC_ERROR_RECOMMENDATION;
  }

  submissionResponse.recommendations = [recommendation];
  const previousResponse = await getSubmissionResponseBySubmissionRequestId(ctx, screeningRequestId, false);
  if (!shouldStoreScreeningResponse(ctx, { submissionRequestId: screeningRequestId, submissionResponse, lastStoredResponse: previousResponse })) {
    logger.info({ ctx, screeningRequestId }, 'Screening request has already a response with the same data, so doing nothing');
  } else {
    await createSubmissionResponse(ctx, submissionResponse);
  }
};
