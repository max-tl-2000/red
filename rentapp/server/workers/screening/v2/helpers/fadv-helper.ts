/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import newId from 'uuid/v4';
import { IApplicantReportRequestTracking, IFadvApplicantData, IRentData, IScreeningResponse } from '../../../../helpers/applicant-types';
import { createApplicantReportRequestTracking } from '../../../../dal/applicant-report-request-tracking-repo';
import { assert } from '../../../../../../common/assert';
import logger from '../../../../../../common/helpers/logger';
import { obscureApplicantProperties } from '../../../../helpers/screening-helper';
import { maskSSNInApplicants } from '../../../../helpers/fadv-mask-applicant-ssn';
import { createFadvRawRequest } from '../../screening-handler-request';
import { IDbContext } from '../../../../../../common/types/base-types';
import { ReportDataBuilder } from './report-data-builder';
import { ApplicantReportNames } from '../../../../../../common/enums/screeningReportTypes';

const createAnApplicantReportRequestTracking = async (
  ctx: IDbContext,
  reportRequestTrackingId: string,
  propertyId: string,
  rentData: IRentData,
  applicantData: any,
  options: any,
): Promise<IApplicantReportRequestTracking> => {
  const { rawRequest } = await createFadvRawRequest(ctx, propertyId, rentData, maskSSNInApplicants(applicantData), options);
  const requestApplicantId = applicantData.applicants[0].applicantId;

  const applicantReportRequestTracking: IApplicantReportRequestTracking = {
    id: reportRequestTrackingId,
    applicantReportId: options.applicantReportId,
    personId: options.personId,
    reportName: options.reportName,
    requestApplicantId,
    propertyId,
    requestType: options.requestType,
    rawRequest,
  };

  return await createApplicantReportRequestTracking(ctx, applicantReportRequestTracking);
};

export const createFadvRequest = async (ctx: IDbContext, propertyId: string, rentData: IRentData, applicantData: IFadvApplicantData, options: any) => {
  assert(rentData && Object.keys(rentData).length > 0, 'createFadvRequest: rentData is missing or empty!');
  logger.debug({ ctx, propertyId, rentData, options, ...obscureApplicantProperties(applicantData) }, 'createAndStoreFadvRequest v2');

  const reportRequestTrackingId = newId();
  const newOptions = {
    ...options,
    submissionRequestId: reportRequestTrackingId,
  };

  const reportRequestResult = await createAnApplicantReportRequestTracking(ctx, reportRequestTrackingId, propertyId, rentData, applicantData, newOptions);
  assert(reportRequestResult.id === reportRequestTrackingId, 'createFadvRequest: the reportRequestTracking id is different');
  logger.debug({ ctx }, 'created submission request');

  const { rawRequest } = await createFadvRawRequest(ctx, propertyId, rentData, applicantData, newOptions);
  logger.debug({ ctx }, 'created raw request');

  logger.debug(
    {
      ctx,
      reportRequestTrackingId,
      propertyId,
      rentData,
      ...obscureApplicantProperties(applicantData),
    },
    'createFadvRequest created request',
  );

  return { rawRequest, id: reportRequestResult.id };
};

export const createReportDataBuilder = (ctx: IDbContext, response: IScreeningResponse, reportLogger?: any) => {
  const reportBuilder = new ReportDataBuilder(ctx, response, reportLogger || logger);
  const buildReport = (reportName: string, applicantId?: string) =>
    reportName === ApplicantReportNames.CREDIT ? reportBuilder.buildCreditReport(applicantId) : reportBuilder.buildCriminalReport(applicantId);
  return {
    buildReport,
    buildCreditReport: (applicantId?: string) => buildReport(ApplicantReportNames.CREDIT, applicantId),
    buildCriminalReport: (applicantId?: string) => buildReport(ApplicantReportNames.CRIMINAL, applicantId),
  };
};
