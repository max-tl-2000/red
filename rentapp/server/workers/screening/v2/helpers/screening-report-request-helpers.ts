/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import newId from 'uuid/v4';
import { ScreeningVersion } from '../../../../../../common/enums/screeningReportTypes';
import { IDbContext, IDictionaryHash } from '../../../../../../common/types/base-types';
import {
  IApplicationData,
  IApplicantReportRequestTracking,
  IScreeningReportOptions,
  IReportRequestData,
  IScreeningReportData,
  IRequestApplicantReportArgs,
  IFadvApplicantData,
  IApplicantReportRequestTrackingWithSettings,
  IApplicantData,
} from '../../../../helpers/applicant-types';
import config from '../../../../../config';
import loggerModule from '../../../../../../common/helpers/logger';
import { getScreeningRequestType, createFadvRawRequest, IFadvRawRequestOptions } from '../../screening-handler-request';
import { FadvRequestTypes } from '../../../../../../common/enums/fadvRequestTypes';
import { APPLICANT_TYPE_APPLICANT } from '../../../../helpers/base-fadv-helper';
import { obscureApplicantProperties } from '../../../../helpers/screening-helper';
import { isPendingRequestTrackingOnTime } from '../../../../helpers/applicant-report-helper';
import { updateApplicantReportRequestTracking } from '../../../../dal/applicant-report-request-tracking-repo';
import { FADV_RESPONSE_STATUS } from '../../../../../common/screening-constants';
import { postRawRequestToScreeningProvider } from '../../screening-provider-integration';
import { getActiveApplicantDataByPersonId } from '../../../../dal/applicant-data-repo';
import { getRequestReportIdfromRaw, getRequestTypefromRaw } from '../../../../../../common/helpers/utils';
import { assert } from '../../../../../../common/assert';

const logger = loggerModule.child({ subType: 'Screening report request helper' });

type GetApplicantsParams = {
  applicationData: IApplicationData;
  personId: string;
  isNotPendingWithoutReportId?: boolean;
};

// TODO: Ask Roberto why ctx is not used
export const getApplicants = (
  _ctx: IDbContext,
  latestApplicantReportRequest: IApplicantReportRequestTracking,
  { applicationData, personId, isNotPendingWithoutReportId }: GetApplicantsParams,
): Array<IApplicationData> => {
  const requestWithSSN: boolean = applicationData.requestWithSSN!;
  const applicantId = !latestApplicantReportRequest || isNotPendingWithoutReportId ? newId() : latestApplicantReportRequest.requestApplicantId;
  const applicant = {
    ...applicationData,
    applicantId,
    personId,
    type: APPLICANT_TYPE_APPLICANT,
    requestWithSSN,
  };

  if ((applicant.ssn || applicant.itin) && applicant.requestWithSSN) applicant.socSecNumber = applicant.ssn || applicant.itin;
  const { ssn, itin, ...otherApplicantInfo } = applicant;

  return [otherApplicantInfo];
};

const handlePendingTimeOutRequest = async (
  ctx: IDbContext,
  isPending: boolean,
  latestApplicantReportRequest: IApplicantReportRequestTracking,
): Promise<void> => {
  if (isPending && !isPendingRequestTrackingOnTime(latestApplicantReportRequest)) {
    assert(latestApplicantReportRequest.id, 'handlePendingTimeOutRequest: Missing latest applicant report request id');

    await updateApplicantReportRequestTracking(ctx, latestApplicantReportRequest.id!, {
      requestEndedAt: new Date(),
      hasTimedOut: true,
    } as IApplicantReportRequestTracking);
  }
};

const isForceNewReportRequested = async (
  ctx: IDbContext,
  latestApplicantReportRequest: IApplicantReportRequestTracking,
  screeningTypeRequested: string,
  isNotPendingWithoutReportId: boolean,
): Promise<boolean> => {
  if (!latestApplicantReportRequest) return true;
  const isPending = !latestApplicantReportRequest.requestEndedAt;

  await handlePendingTimeOutRequest(ctx, isPending, latestApplicantReportRequest);
  const isPendingWithNewStatus = isPending && latestApplicantReportRequest.requestType === FadvRequestTypes.NEW;
  return screeningTypeRequested === FadvRequestTypes.NEW || isNotPendingWithoutReportId || isPendingWithNewStatus;
};

const getScreeningReportOptions = async (
  ctx: IDbContext,
  latestApplicantReportRequest: IApplicantReportRequestTracking,
  screeningTypeRequested: string,
  isNotPendingWithoutReportId: boolean,
): Promise<IScreeningReportOptions> => {
  const screeningReportOptions = {
    storeRequest: true,
    version: ScreeningVersion.V2,
    requestType: getScreeningRequestType(screeningTypeRequested, FadvRequestTypes.NEW),
  };
  const forceNewReportRequested = await isForceNewReportRequested(ctx, latestApplicantReportRequest, screeningTypeRequested, isNotPendingWithoutReportId);
  if (forceNewReportRequested) {
    logger.trace({ ctx, screeningReportOptions }, 'forcing a new screening');
    return {
      ...screeningReportOptions,
      requestType: FadvRequestTypes.NEW,
    };
  }

  return {
    ...screeningReportOptions,
    reportId: latestApplicantReportRequest.externalReportId,
    requestType: getScreeningRequestType(screeningTypeRequested, FadvRequestTypes.MODIFY),
  };
};

export const getReportRequestData = async (
  ctx: IDbContext,
  latestApplicantReportRequest: IApplicantReportRequestTracking,
  // TODO: Ask Ivonne why applicantReportId was not used
  { /* applicantReportId, */ screeningTypeRequested, personId, reportName, applicationData, propertyId }: IRequestApplicantReportArgs,
): Promise<IReportRequestData> => {
  const { rentData } = config.fadv;
  logger.trace({ ctx, personId, reportName }, 'getReportRequestData');

  const isNotPendingWithoutReportId =
    latestApplicantReportRequest && latestApplicantReportRequest.requestEndedAt && !latestApplicantReportRequest.externalReportId;
  const screeningReportOptions: IScreeningReportOptions = await getScreeningReportOptions(
    ctx,
    latestApplicantReportRequest,
    screeningTypeRequested!,
    isNotPendingWithoutReportId!,
  );

  const applicantData: IFadvApplicantData = {
    tenantId: ctx.tenantId,
    applicants: getApplicants(ctx, latestApplicantReportRequest, {
      applicationData,
      personId,
      isNotPendingWithoutReportId,
    }),
  };

  const screeningReportData: IScreeningReportData = {
    rentData,
    propertyId,
    applicantData,
  };

  logger.trace(obscureApplicantProperties(screeningReportData), 'got screening report data');

  return { screeningReportData, screeningReportOptions };
};

export const handleImmediateScreeningResponse = async (
  ctx: IDbContext,
  { response, screeningRequestId }: { response: IDictionaryHash<any>; screeningRequestId: string },
): Promise<void> => {
  const responseNode = response.ApplicantScreening.Response[0];
  const status = responseNode.Status[0];
  logger.trace({ ctx, screeningRequestId, responseStatus: status }, 'handleImmediateScreeningResponse');

  const hadScreeningResult = status === FADV_RESPONSE_STATUS.COMPLETE;
  const fadvStatuses = [FADV_RESPONSE_STATUS.COMPLETE.toLowerCase(), FADV_RESPONSE_STATUS.INCOMPLETE.toLowerCase()];
  const externalReportId = responseNode.TransactionNumber && responseNode.TransactionNumber[0];
  if (fadvStatuses.includes(status.toLowerCase()) && externalReportId) {
    logger.trace(
      { ctx, screeningRequestId, responseStatus: status, hadScreeningResult },
      'handleImmediateScreeningResponse updating request with transactionNumber',
    );
    await updateApplicantReportRequestTracking(ctx, screeningRequestId!, {
      externalReportId,
      requestEndedAt: new Date(),
    } as IApplicantReportRequestTracking);
  } else {
    logger.error({ ctx, screeningRequestId }, 'missing transaction number on screening response');
  }

  const { handleScreeningResponseReceived } = require('../../screening-handler-response'); // eslint-disable-line global-require
  try {
    logger.trace({ ctx, screeningRequestId }, 'handleImmediateScreeningResponse calling responseReceived handler');
    await handleScreeningResponseReceived({ ...ctx, screeningResponse: response });
  } catch (err) {
    logger.error({ ctx, screeningRequestId, err }, 'unable to process immediate response!');
    throw err;
  }
};

export const prepareRawRequestForFadvResubmit = async (
  ctx: IDbContext,
  screeningRequest: IApplicantReportRequestTrackingWithSettings,
  applicantData: IFadvApplicantData,
) => {
  const rawRequestReportId = getRequestReportIdfromRaw(screeningRequest);
  const rawRequestType = getRequestTypefromRaw(screeningRequest);
  const { reportName } = screeningRequest;

  const options: IFadvRawRequestOptions = {
    submissionRequestId: screeningRequest.id as string,
    reportId: rawRequestReportId,
    requestType: rawRequestType,
    version: ScreeningVersion.V2,
    reportName,
  };

  const { rentData } = config.fadv;

  const { rawRequest } = await createFadvRawRequest(ctx, screeningRequest.propertyId, rentData, applicantData, options);

  return rawRequest;
};

export const handleScreeningResubmitRequest = async (ctx: IDbContext, screeningRequest: IApplicantReportRequestTrackingWithSettings) => {
  const screeningRequestId = screeningRequest.id!;
  logger.debug({ ctx, screeningRequestId }, 'handleScreeningResubmitRequest');
  const applicationData: IApplicationData = ((await getActiveApplicantDataByPersonId(ctx, screeningRequest.personId!)) || ({} as IApplicantData))
    .applicationData;

  const applicantData: IFadvApplicantData = {
    tenantId: ctx.tenantId,
    applicants: getApplicants(ctx, screeningRequest, {
      applicationData,
      personId: screeningRequest.personId!,
    }),
  };

  try {
    const rawRequest = await prepareRawRequestForFadvResubmit(ctx, screeningRequest, applicantData);

    const { response } = await postRawRequestToScreeningProvider(ctx, {
      screeningRequestId,
      payload: rawRequest,
    });
    await handleImmediateScreeningResponse(ctx, {
      response,
      screeningRequestId,
    });
  } catch (err) {
    logger.error({ err, ctx, screeningRequestId }, 'Unable to post message to screening provider');
    throw new Error('Unable to post message to screening provider');
  }
};
