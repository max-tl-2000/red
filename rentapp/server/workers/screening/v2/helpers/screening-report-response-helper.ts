/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import { IParsedResponse } from '../../../../helpers/applicant-types';
import { IDictionaryHash } from '../../../../../../common/types/base-types';
import { BlockedServiceErrors } from '../../../../services/helpers/party-application-helper';
import { ServiceNames, REVA_SERVICE_STATUS } from '../../../../../../common/enums/applicationTypes';
import { ApplicantReportNames } from '../../../../../../common/enums/screeningReportTypes';

const ADDRESS_ERROR_CODES = ['30011', '30017', '30022'];
const EXPIRED_ERROR_CODE = '30007';
export const DEFAULT_GRACE_PERIOD = 30;

export const isAddressError = (errorCode: string): boolean => ADDRESS_ERROR_CODES.includes(errorCode);
export const isExpiredError = (errorCode: string): boolean => errorCode === EXPIRED_ERROR_CODE;
export const isStatusBlocked = (status: string): boolean => status === REVA_SERVICE_STATUS.BLOCKED;
export const isCreditFreeze = (response: IParsedResponse | IDictionaryHash<any> = {}): boolean => {
  const { serviceStatus, BlockedStatus } = response;
  if (!serviceStatus || !BlockedStatus) return false;

  const isCreditStatusBlocked = serviceStatus[ServiceNames.CREDIT] ? isStatusBlocked(serviceStatus[ServiceNames.CREDIT].status) : false;
  return isCreditStatusBlocked && BlockedStatus.includes(BlockedServiceErrors.CREDIT_FREEZE);
};

export const GRACE_PERIOD_MAPPER = {
  [ApplicantReportNames.CREDIT]: 'creditReportValidForPeriod',
  [ApplicantReportNames.CRIMINAL]: 'criminalReportValidForPeriod',
};
