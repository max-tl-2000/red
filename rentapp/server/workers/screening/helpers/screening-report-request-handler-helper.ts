/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import omit from 'lodash/omit';
import loggerModule from '../../../../../common/helpers/logger';
import { getScreeningRequest } from '../../../services/screening';
import { NoRetryError } from '../../../../../server/common/errors';
import { obscureApplicantProperties } from '../../../helpers/screening-helper';
import { IRequestApplicantReportArgs, IScreeningReportOptions, ScreeningResponseOrigin } from '../../../helpers/applicant-types';
import { FadvRequestTypes } from '../../../../../common/enums/fadvRequestTypes';
import { ScreeningVersion } from '../../../../../common/enums/screeningReportTypes';
import { postToScreeningProvider, processScreeningResponseReceived } from '../screening-provider-integration-handler';
import { getReportRequestData } from '../v2/helpers/screening-report-request-helpers';
import { SCREENING_MESSAGE_TYPE } from '../../../../../server/helpers/message-constants';
import { IDictionaryHash, IConsumerResult, IDbContext } from '../../../../../common/types/base-types';
import { getApplicantReportRequestWithApplicantData } from '../../../dal/applicant-report-request-tracking-repo';
import { handleFADVResponseError } from './screening-report-response-handler-helper';
import { ScreeningDecision } from '../../../../../common/enums/applicationTypes';
import { PARSE_EXCEPTION } from '../../../../../common/helpers/postXML';
import { PARSE_ERROR } from '../../../../common/screening-constants';

const logger = loggerModule.child({ subType: 'Application Screening Handler' });
const decryptSsn = true;

const getOrphanedScreeningRequest = async (
  ctx: IDbContext,
  screeningData: { screeningRequestId: string; personId?: string; reportName?: string; propertyId?: string; reportId?: string },
): Promise<IDictionaryHash<any>> => {
  const { screeningVersion } = ctx;
  const { screeningRequestId, personId, reportName, propertyId, reportId } = screeningData;

  if (!screeningVersion || screeningVersion === ScreeningVersion.V1) {
    return await getScreeningRequest(ctx, screeningRequestId, decryptSsn);
  }

  let latestApplicantReportRequest;
  const screeningRequest = await getApplicantReportRequestWithApplicantData(ctx, screeningRequestId);
  const { screeningReportData } = await getReportRequestData(ctx, latestApplicantReportRequest, {
    applicantReportId: reportId,
    screeningTypeRequested: FadvRequestTypes.VIEW,
    personId,
    reportName,
    applicationData: screeningRequest.applicantData,
    propertyId,
  } as IRequestApplicantReportArgs);

  const { rentData, applicantData } = screeningReportData;

  return {
    ...screeningRequest,
    rentData,
    applicantData,
    reportName,
  };
};

export const handleScreeningSubmitViewRequestReceived = async (msg: IDictionaryHash<any>): Promise<IConsumerResult> => {
  const { screeningRequestId, partyApplicationId, ...ctx } = msg;
  const handlerCtx = { ...ctx } as IDbContext;
  const { tenantId, screeningVersion: version = ScreeningVersion.V1, reportId, personId, reportName, propertyId } = handlerCtx;
  const responseOrigin = ScreeningResponseOrigin.POLL;

  logger.debug({ ctx: handlerCtx, screeningRequestId, screeningVersion: version }, 'handleScreeningSubmitViewRequestReceived');
  if (!tenantId) throw new NoRetryError('msg was missing tenantId');
  if (!screeningRequestId) throw new NoRetryError('msg was missing screeningRequestId');

  const screeningRequest = await getOrphanedScreeningRequest(handlerCtx, { screeningRequestId, personId, reportName, propertyId, reportId });

  logger.info({ ctx: handlerCtx, partyApplicationId, screeningRequestId }, 'handleScreeningSubmitViewRequestReceived');

  try {
    const applicantData = await screeningRequest.applicantData;
    const options: IScreeningReportOptions = {
      storeRequest: false,
      requestType: FadvRequestTypes.VIEW,
      reportId: screeningRequest.transactionNumber,
      submissionRequestId: screeningRequestId,
      reportName,
      version,
    };

    const eventType = SCREENING_MESSAGE_TYPE.SCREENING_SUBMIT_VIEW_REQUEST_RECEIVED;
    const { response } = await postToScreeningProvider(
      handlerCtx,
      screeningRequest.propertyId,
      { ...screeningRequest.rentData, quoteId: screeningRequest.quoteId },
      applicantData,
      { ...options, eventType },
    );
    const fadvResponseLog = omit(response.ApplicantScreening.Response[0], ['BackgroundReport']);
    logger.info({ ctx: handlerCtx, ...obscureApplicantProperties(fadvResponseLog) }, 'postToScreeningProvider got response');

    const fadvResponse = response.ApplicantScreening.Response[0];
    const status = fadvResponse.Status[0];
    logger.info({ ctx: handlerCtx, fadvResponseStatus: status }, 'Sending response from FADV as a posted notification');
    await processScreeningResponseReceived(handlerCtx, response, { version, responseOrigin });
  } catch (err) {
    logger.error({ err, ctx: handlerCtx, screeningRequestId }, 'Unable to post message to screening provider');

    if (err.msg === PARSE_EXCEPTION || err.token === PARSE_ERROR) {
      await handleFADVResponseError(ctx, err.data, ScreeningDecision.ERROR_RESPONSE_UNPARSABLE, {
        screeningRequestId: err.screeningRequestId,
        responseOrigin,
      });
      return { processed: false } as IConsumerResult;
    }
    throw new Error('Unable to post message to screening provider');
  }

  return { processed: true } as IConsumerResult;
};
