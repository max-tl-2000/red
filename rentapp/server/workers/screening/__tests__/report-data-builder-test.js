/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import newId from 'uuid/v4';
import path from 'path';
import loggerModule from '../../../../../common/helpers/logger';
import { createReportDataBuilder } from '../v2/helpers/fadv-helper';
import { readJSON } from '../../../../../common/helpers/xfs';

describe('Report Data Builder', () => {
  const getParsedFadvResponse = async (fileName = 'v2-fadv-response.json') =>
    await readJSON(path.resolve(path.dirname(__dirname), '__tests__', 'resources', fileName));

  const ctx = { tenantId: newId(), trx: newId() };
  const logger = loggerModule.child({ subType: 'Screening v2 Report Data Builder', ctx });

  describe('When not parsedFadvResponse is passed in the constructor', () => {
    it('should throw an error', () => {
      expect(() => createReportDataBuilder(ctx)).toThrow('fadvResponse not defined');
    });
  });

  describe('When building a Credit Report data object', () => {
    it('should return an object containing the correct keys and values', async () => {
      const screeningResponse = await getParsedFadvResponse();
      const reportBuilder = createReportDataBuilder(ctx, screeningResponse, logger);
      const reportData = reportBuilder.buildCreditReport('30072231');
      expect(reportData).toMatchSnapshot();
    });

    it('and is a requried ssn reponse, should return an object with hasRequiredSsnResponse set to true ', async () => {
      const screeningResponse = await getParsedFadvResponse('required-ssn-response.json');
      const reportBuilder = createReportDataBuilder(ctx, screeningResponse, logger);
      const { hasRequiredSsnResponse } = reportBuilder.buildCreditReport('31719437');

      expect(hasRequiredSsnResponse).toEqual(true);
    });
  });

  describe('When building a Criminal Report data object', () => {
    it('should return an object containing the correct keys and values', async () => {
      const screeningResponse = await getParsedFadvResponse();
      const reportBuilder = createReportDataBuilder(ctx, screeningResponse, logger);
      const reportDataA = reportBuilder.buildCriminalReport('30072231');
      const reportDataB = reportBuilder.buildCriminalReport('30072232');

      expect(reportDataA).toMatchSnapshot();
      // EV702 F, SW852 F,
      expect(reportDataA.hasEvictionFilingsMin).toEqual(true);
      // EV702 P, SW852 P,
      expect(reportDataB.hasEvictionFilingsMin).toEqual(false);
    });
  });
});
