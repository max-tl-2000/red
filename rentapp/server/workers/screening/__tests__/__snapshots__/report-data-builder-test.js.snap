// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Data Builder When building a Credit Report data object should return an object containing the correct keys and values 1`] = `
Object {
  "creditScore": 736,
  "hasBankruptcy": false,
  "hasEvictionNotice": false,
  "hasForeclosure": false,
  "hasLeaseViolation": false,
  "hasLegalItems": false,
  "hasMortgageDebt": false,
  "hasNoEstablishedCredit": true,
  "hasNsfOrLatePayMax": null,
  "hasNsfOrLatePayMin": false,
  "hasPropertyRentalDebt": false,
  "hasRentalCollections": false,
  "hasRequiredSsnResponse": null,
  "hasSuspiciousSsn": false,
  "hasTaxLiens": false,
  "hasUtilityDebt": false,
}
`;

exports[`Report Data Builder When building a Criminal Report data object should return an object containing the correct keys and values 1`] = `
Object {
  "hasDrugsFelony": false,
  "hasDrugsMisdemeanor": false,
  "hasDuiFelony": false,
  "hasDuiMisdemeanor": null,
  "hasEvictionFilingsMax": false,
  "hasEvictionFilingsMin": true,
  "hasEvictionRecordsMax": false,
  "hasEvictionRecordsMin": null,
  "hasGlobalSanctions": null,
  "hasGlobalSanctionsFuzzy": null,
  "hasPropertyFelony": false,
  "hasPropertyMisdemeanor": false,
  "hasRegisteredSexOffender": false,
  "hasRegisteredSexOffenderDobOnly": false,
  "hasSexFelony": false,
  "hasSexMisdemeanor": false,
  "hasTheftByCheckFelony": false,
  "hasTheftByCheckMisdemeanor": false,
  "hasTheftFelony": false,
  "hasTheftMisdemeanor": false,
  "hasTrafficFelony": null,
  "hasTrafficMisdemeanor": null,
  "hasUnclassifiedFelony": false,
  "hasUnclassifiedMisdemeanor": false,
  "hasViolentCrimeFelony": false,
  "hasViolentCrimeMisdemeanor": false,
  "hasWeaponsFelony": false,
  "hasWeaponsMisdemeanor": false,
}
`;
